proposals/0052-svg-components.md

```diff
@@ -0,0 +1,247 @@
+
+- Start Date: 2024-10-01
+- Reference Issues: <!-- related issues, otherwise leave empty -->
+- Implementation PR: https://github.com/withastro/astro/pull/12067
+- Stage 2 Issue: https://github.com/withastro/roadmap/issues/699
+- Stage 3 PR: <!-- related roadmap PR, leave it empty if you don't have a PR yet -->
+
+# Summary
+
+This RFC proposes adding native support for importing and rendering `.svg` files as Astro components, with optimized rendering techniques to minimize performance impact. It aims to allow `.svg` files to be treated as components that accept props and be optimized for repeated use on a page using `<symbol>` and `<use>` elements.
+
+# Example
+
+An SVG can be imported directly into an Astro component and used as a component that will only embed itself once. 
+
+```astro
+---
+import Logo from '../assets/logo.svg'
+---
+
+<!-- First usage: includes initial <symbol> -->
+<Logo width={48} height={48} />
+
+<!-- Subsequent usage(s): automatically optimized with <use> -->
+<Logo width={48} height={48} />
+```
+
+# Background & Motivation
+
+Include any useful background detail that that explains why this RFC is important.
+What are the problems that this RFC sets out to solve? Why now? Be brief!
+
+It can be useful to illustrate your RFC as a user problem in this section.
+(ex: "Users have reported that it is difficult to do X in Astro today.")
+
+# Goals
+
+- Enable importing and rendering `.svg` files as Astro components.
+- Allow users to pass props to the root `svg` element while allowing for prop overrides.
+- Inline `.svg` file directly without using optimized sprite rendering behavior.
+- Allow users to opt-in to optimized `.svg` rendering by utilizing `symbol` and `use` for repeated instances.
+- Maintain backwards compatibility with existing `.svg` import behavior.
+- Strip unnecessary attributes like `xmlns` and `version` from `.svg` files to reduce size.
+- Allow third-party plugins to hook into the `.svg` import behavior to generate their own component libraries.
+
+# Non-Goals
+
+- Out-of-scope: Advanced tree-shaking or SSR optimizations for large icon sets.
+- Out-of-scope: Advanced `.svg` file optimization such as `svgo`.
+- Future: Adding support for `.svg` files as framework components.
+- Non-goal: Adding support for `.json` icon sets such as `@iconify/json`.
+- Non-goal: SVG content analysis for accessibility.
+
+# Detailed Design
+
+## Importing SVG Files as Components
+
+The core design of this proposal allows `.svg` files to be imported and used as components in Astro. This means developers can import `.svg` files into their Astro components and render them directly as if they were Astro components. The goal is to maintain ease of use while automating performance optimizations under the hood.
+
+**Example:**
+
+```astro
+---
+import Logo from '../assets/logo.svg';
+---
+
+<Logo width={48} height={48} />
```

Comment:
All the examples use both `width` and `height`. I believe we should also explain if both props are mandatory. If not, we should explain the end result. If so, we should explain if we plan to raise an error.

Reply:
Yes, I agree. This is a bit of copy/paste late at night before submitting. I'll clean these examples up so it doesn't hint that height/width are required and demonstrates more of the intended examples.